File: template-service/src/main/java/com/sensa/templateservice/service/TemplateService.java
package com.sensa.templateservice.service;

import com.sensa.templateservice.dto.TemplateRequest;
import com.sensa.templateservice.dto.TemplateResponse;
import com.sensa.templateservice.exception.TemplateAlreadyExistException;
import com.sensa.templateservice.exception.TemplateNotFoundException;
import com.sensa.templateservice.mapper.TemplateMapper;
import com.sensa.templateservice.repository.TemplateRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
public class TemplateService {
    private final TemplateRepository templateRepository;
    private final TemplateMapper templateMapper;

    public TemplateResponse createTemplate(String username, TemplateRequest templateRequest) {
        return Optional.of(templateRepository.existTemplateByUsernameAndTitle(username, templateRequest.title()))
                .filter(exist -> !exist)
                .map(unused -> templateMapper.mapToEntity(templateRequest))
                .map(template -> template.addUsername(username))
                .map(templateRepository::save)
                .map(templateMapper::mapToResponse)
                .orElseThrow(() -> new TemplateAlreadyExistException(
                        String.format("Template with title %s already exists for user: %s", templateRequest.title(), username)
                ));
    }

    public TemplateResponse findTemplate(String username, String title) {
        return templateRepository.findByClientUsernameAndTitle(username, title)
                .map(templateMapper::mapToResponse)
                .orElseThrow(() -> new TemplateNotFoundException(
                   String.format("Template with title %s not found", title)
                ));
    }

    public TemplateResponse updateTemplate(
            String username, String title,
            TemplateRequest templateRequest) {
        return templateRepository.findByClientUsernameAndTitle(username, title)
                .map(existingTemplate -> {
                    Optional.ofNullable(templateRequest.title()).ifPresent(existingTemplate::setTitle);
                    Optional.ofNullable(templateRequest.content()).ifPresent(existingTemplate::setContent);
                    return existingTemplate;
                })
                .map(templateRepository::save)
                .map(templateMapper::mapToResponse)
                .orElseThrow(() -> new TemplateNotFoundException(
                        String.format("Template with title %s not found for user: %s", templateRequest.title(), username)
                ));
    }

    public Boolean deleteTemplate(String username, String title) {
        int rowsDeleted = templateRepository.deleteByClientUsernameAndTitle(username, title);
        return rowsDeleted != 0;
    }
}

File: template-service/src/main/kotlin/com/sensa/templateservice/service/TemplateService.kt
package com.sensa.templateservice.service

import com.sensa.templateservice.dto.TemplateRequest
import com.sensa.templateservice.dto.TemplateResponse
import com.sensa.templateservice.entity.TemplateChannel
import com.sensa.templateservice.exception.TemplateAlreadyExistsException
import com.sensa.templateservice.exception.TemplateNotFoundException
import com.sensa.templateservice.mapper.TemplateMapper
import com.sensa.templateservice.repository.TemplateRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
class TemplateService(
    private val templateRepository: TemplateRepository,
    private val templateMapper: TemplateMapper
) {
    @Transactional
    fun create(request: TemplateRequest, createdBy: UUID): TemplateResponse {
        if (templateRepository.existsByTemplateName(request.templateName)) {
            throw TemplateAlreadyExistsException(request.templateName)
        }
        val entity = templateMapper.toEntity(request, createdBy)
        return templateMapper.toResponse(templateRepository.save(entity))
    }

    fun getByTemplateName(templateName: String): TemplateResponse {
        val entity = templateRepository.findByTemplateName(templateName)
            ?: throw TemplateNotFoundException(templateName)
        return templateMapper.toResponse(entity)
    }

    fun getAll(): List<TemplateResponse> {
        return templateRepository.findAll().map { templateMapper.toResponse(it) }
    }

    fun getByChannel(channel: TemplateChannel): List<TemplateResponse> {
        return templateRepository.findByChannel(channel).map { templateMapper.toResponse(it) }
    }

    @Transactional
    fun update(templateName: String, request: TemplateRequest): TemplateResponse {
        val existing = templateRepository.findByTemplateName(templateName)
            ?: throw TemplateNotFoundException(templateName)
        templateMapper.updateEntity(existing, request)
        return templateMapper.toResponse(templateRepository.save(existing))
    }

    @Transactional
    fun delete(templateName: String) {
        val rows = templateRepository.deleteByTemplateName(templateName)
        if (rows == 0) throw TemplateNotFoundException(templateName)
    }
}
